File: client/typescript/gamium-client-common/src/ui/ui.ts
import { Vector3T } from "../protocols/generated";
import { GamiumClient } from "../gamium-client";
import { GamiumService } from "../gamium-service";
import { isLocator, Locator } from "../locator/locator";
import { UIElement } from "../object/ui-element";
import {
  ActionsClickOptions,
  DefaultActionClickOptions,
} from "../options/actions-click-options";
import {
  ActionsDragOptions,
  DefaultActionsDragOptions,
} from "../options/actions-drag-options";
import {
  ActionsScrollOptions,
  DefaultActionScrollOptions,
} from "../options/actions-scroll-options";
import {
  DefaultFindObjectOptions,
  FindObjectOptions,
} from "../options/find-object-options";
import { Vector2 } from "../protocols/types";
import { tryify, TryResult } from "../try";
import { Until } from "../until";

export class UI {
  constructor(
    protected readonly game: GamiumClient,
    protected readonly gamiumService: GamiumService
  ) {}

  async find(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<UIElement> {
    const info = await this.game.wait(Until.objectLocated(locator, options));
    return new UIElement(this.game, this.gamiumService, info);
  }

  async tryFind(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<TryResult<UIElement>> {
    return tryify(this.find(locator, options));
  }

  async finds(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<UIElement[]> {
    const infos = await this.game.wait(Until.objectsLocated(locator, options));
    const elements = infos.map(
      (i) => new UIElement(this.game, this.gamiumService, i)
    );
    return elements;
  }

  async tryFinds(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<TryResult<UIElement[]>> {
    return tryify(this.finds(locator, options));
  }

  async click(
    locator: Locator,
    options: Partial<FindObjectOptions & ActionsClickOptions> = {
      ...DefaultFindObjectOptions(),
      ...DefaultActionClickOptions(),
    }
  ): Promise<void> {
    const element = await this.find(locator, options);
    await this.game.wait(Until.elementInteractable(element));
    await element.click(options);
  }

  async drag(
    locator: Locator,
    to: Locator | Vector2,
    options: Partial<FindObjectOptions & ActionsDragOptions> = {
      ...DefaultFindObjectOptions(),
      ...DefaultActionsDragOptions(),
    }
  ): Promise<void> {
    const element = await this.find(locator, options);
    await this.game.wait(Until.elementInteractable(element));
    let toPos = new Vector3T(0, 0, 0);
    if (isLocator(to)) {
      const toElement = await this.find(to, options);
      toPos = new Vector3T(
        toElement.info.screenPosition.x,
        toElement.info.screenPosition.y,
        toElement.info.screenPosition.z
      );
    } else {
      toPos = new Vector3T(to.x, to.y, 0);
    }
    await this.game
      .actions()
      .drag(element.info.screenPosition, toPos, options)
      .perform();
  }

  async scroll(
    locator: Locator,
    delta: Vector2,
    options: Partial<FindObjectOptions & ActionsScrollOptions> = {
      ...DefaultFindObjectOptions(),
      ...DefaultActionScrollOptions(),
    }
  ): Promise<void> {
    const element = await this.find(locator, options);
    await this.game.wait(Until.elementInteractable(element));
    await element.scroll(delta, options);
  }

  async setText(
    locator: Locator,
    text: string,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<void> {
    const element = await this.find(locator, options);
    await this.game.wait(Until.elementInteractable(element));
    await element.setText(text);
  }

  async getText(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<string> {
    const element = await this.find(locator, options);
    await this.game.wait(Until.elementInteractable(element));
    return await element.getText();
  }
}

File: client/python/gamium/gamium/ui/ui.py
import functools
from typing import List, Optional, Union
from gamium.protocol.generated.Types.Vector2 import Vector2T
from gamium.protocol.generated.Types.Vector3 import Vector3T
from gamium.condition.until import Until
from gamium.object.ui_element import UIElement
from gamium.options.action_click_options import ActionClickOptions
from gamium.options.action_drag_options import ActionDragOptions
from gamium.options.action_scroll_options import ActionScrollOptions
from gamium.protocol.types import ObjectInfo, Vector2
from gamium.igamium_client import IGamiumClient
from gamium.igamium_service import IGamiumService
from gamium.internal.logger import Logger
from gamium.locator.locator import Locator
from gamium.options.find_objects_options import FindObjectOptions
from gamium.utils.try_utils import TryResult, tryify


class UI:
    def __init__(self, client: IGamiumClient, service: IGamiumService):
        self._client = client
        self._service = service

    def find(self, locator: Locator, options: FindObjectOptions = FindObjectOptions()) -> UIElement:
        info = self._client.wait(Until.object_located(locator, options))
        return UIElement(self._client, self._service, info)

    def try_find(self, locator: Locator, options: FindObjectOptions = FindObjectOptions()) -> TryResult[UIElement]:
        return tryify(functools.partial(self.find, locator, options))

    def finds(self, locator: Locator, options: FindObjectOptions = FindObjectOptions()) -> List[UIElement]:
        infos = self._client.wait(Until.objects_located(locator, options))
        elements = []
        for info in infos:
            elements.append(UIElement(self._client, self._service, info))
        return elements

    def try_finds(self, locator: Locator, options: FindObjectOptions = FindObjectOptions()) -> TryResult[List[UIElement]]:
        return tryify(functools.partial(self.finds, locator, options))

    def click(
        self,
        locator: Locator,
        find_options: FindObjectOptions = FindObjectOptions(),
        click_options: ActionClickOptions = ActionClickOptions(),
    ):
        elem = self.find(locator, find_options)
        self._client.wait(Until.element_interactable(elem))
        elem.click(click_options)

    def drag(
        self,
        locator: Locator,
        to: Union[Locator, Vector2],
        find_options: FindObjectOptions = FindObjectOptions(),
        drag_options: ActionDragOptions = ActionDragOptions(),
    ):
        elem = self.find(locator, find_options)
        self._client.wait(Until.element_interactable(elem))
        to_pos = Vector2.zero()
        if isinstance(to, Locator):
            to_elem = self.find(to)
            to_pos = Vector2.from_vector3(to_elem.info.screen_position)
        if isinstance(to, Vector2):
            to_pos = to
        elem_pos = Vector2.from_vector3(elem.info.screen_position)
        self._client.actions().drag(elem_pos, to_pos, drag_options).perform()

    def scroll(
        self,
        locator: Locator,
        delta: Vector2,
        find_options: FindObjectOptions = FindObjectOptions(),
        scroll_options: ActionScrollOptions = ActionScrollOptions(),
    ):
        elem = self.find(locator, find_options)
        self._client.wait(Until.element_interactable(elem))
        elem.scroll(delta, scroll_options)

    def set_text(
        self,
        locator: Locator,
        text: str,
        find_options: FindObjectOptions = FindObjectOptions(),
    ):
        elem = self.find(locator, find_options)
        self._client.wait(Until.element_interactable(elem))
        elem.set_text(text)

    def get_text(
        self,
        locator: Locator,
        find_options: FindObjectOptions = FindObjectOptions(),
    ):
        elem = self.find(locator, find_options)
        self._client.wait(Until.element_interactable(elem))
        return elem.get_text()
